File: java/accumulate/src/main/java/Accumulate.java
import java.util.ArrayList;
import java.util.List;
import java.util.function.UnaryOperator;

public class Accumulate {
    public static <T> List<T> accumulate(List<T> xs, UnaryOperator<T> f) {
        return xs.stream()
                .collect(ArrayList::new,
                        (acc, x) -> acc.add(f.apply(x)),
                        ArrayList::addAll);
    }
}

File: haskell/accumulate/src/Accumulate.hs
module Accumulate (accumulate) where

accumulate :: (a -> b) -> [a] -> [b]

-- accumulate _ [] = []
-- accumulate fn (x:xs) = fn x : accumulate fn xs

-- accumulate fn xs = foldr (\x acc-> fn x:acc) [] xs

accumulate fn = foldr ((:) . fn) []